# **Лабораторная работа №1. Обработка данных учебной аналитики с использованием методов машинного обучения**

# Описание работы

**Цель работы:** сформировать навыки обработки данных учебной аналитики с использованием методов машинного обучения. </br>
**Задачи:**


1.   Изучить предоставленный набор данных.
2.   Собрать данные согласно своему варианту.
2.   Выполнить предобработку данных.
2.   Изучить алгоритмы машинного обучения для решения задачи классификации.
2.   Обосновать, для каких алгоритмов необходимо сделать бинаризацию и нормализацию признаков.
2.   Построить модели классификации.
2.   Вычислить метрики качества моделей.
3.   С помощью подбора параметров добиться точности не менее 0.7 (для логистической регрессии - не менее 0.6).
3.   Сделать выводы по работе.

**Формат сдачи работы** - отчет в формате PDF, оформленный по актуальному ГОСТу по оформлению ВКР в Университете ИТМО, и оформленный Google Colab по аналогии с практической работой.

# 1-2 Загрузка текста и изучение

In [ ]:
# Вычисление варианта
student = 15 # напишите свой номер в списке групп
print('Мой вариант студенты номер - ', student, student*2, student*3-1)

Мой вариант студенты номер -  15 30 44


In [ ]:
# Импортируем основную библиотеку для работы с данными
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#@Подгружаем данные о 15 и 30 и 44 студенте
data_15_1 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 1/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_1 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 1/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_1 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 1/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])

data_15_2 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 2/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_2 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 2/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_2 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 2/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])

data_15_3 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 3/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_3 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 3/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_3 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 3/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])

data_15_4 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 4/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_4 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 4/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_4 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 4/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])

data_15_5 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 5/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_5 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 5/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_5 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 5/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])

data_15_6 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 6/15', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_30_6 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 6/30', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_6 = pd.read_csv('/content/drive/MyDrive/New_stuff/7-s sem/7 - Training analyst/EPM Dataset 2/Data/Processes/Session 6/44', 
                      names=['session', 'student_Id', 'exercise', 'activity', 'start_time', 'end_time', 'idle_time', 'mouse_wheel', 
                             'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke'])
data_44_6.shape

(687, 13)

In [ ]:
data_15_6.head()

session  student_Id exercise  ... mouse_click_right mouse_movement keystroke
0        6          15       Es  ...                 0             81         0
1        6          15       Es  ...                 0             50         0
2        6          15   Es_6_1  ...                 0            138         0
3        6          15   Es_6_1  ...                 0            115         0
4        6          15   Es_6_1  ...                 0            173         0

[5 rows x 13 columns]

# 3 Обработка

In [ ]:
#Объединяем в один датафрейм
data_full = pd.concat([data_15_1, data_15_2, 
                     data_15_3, data_15_4, 
                     data_15_5, data_15_6,
                     data_30_1, data_30_2, 
                     data_30_3, data_30_4, 
                     data_30_5, data_30_6,
                      data_44_1, data_44_2, 
                     data_44_3, data_44_4, 
                     data_44_5, data_44_6],
                    sort=False, axis=0)
data_full.shape

(8938, 13)

In [ ]:
#Выбираем типы ACTiVITY с которыми будем работать
data_full.activity.value_counts()

 Other                1441
 Diagram               796
 Blank                 726
 Properties            597
 Aulaweb               354
                      ... 
 Deeds_Es                4
 TextEditor_Es_6_4       4
 TextEditor_Es           2
 Fsm_Related             1
 Study_Es_6_6            1
Name: activity, Length: 95, dtype: int64

In [ ]:
#Посмотрим, как записываются типы в датафрейме
data_full.activity.values[:10]

array([' Other', ' Deeds', ' Study_Es_1_1', ' Other', ' Study_Es_1_1',
       ' Deeds_Es_1_1', ' Deeds_Es_1_1', ' Other', ' Deeds_Es_1_1',
       ' Study_Es_1_1'], dtype=object)

In [ ]:
# Создаем новый датафрейм, который содержит в себе записи только с двумя типами действия
new_data_full = data_full.loc[data_full['activity'].isin([' Diagram', ' Other'])]
new_data_full.shape

(2237, 13)

# 4-6 Изучение алгоритмов и применение, обоснование бинаризация и нормализации признаков

In [ ]:
# Выделяем целевую переменную и признаки.
y = new_data_full['activity']
y.shape

(2237,)

In [ ]:
# Выбираем только численные признаки
features = ['idle_time', 'mouse_wheel', 'mouse_wheel_click', 'mouse_click_left', 'mouse_click_right', 'mouse_movement', 'keystroke']
X = new_data_full[features]
X.shape

(2237, 7)

In [ ]:
# Разбиваем выборку на обучающую и тестовую
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4, random_state = 1)

In [ ]:
X_train

idle_time  mouse_wheel  ...  mouse_movement  keystroke
99        2171            0  ...             178          0
107      47700            0  ...            1027          0
16          32            1  ...             562          0
355          0            0  ...              57          0
501          0            0  ...             323          0
..         ...          ...  ...             ...        ...
25      235248            0  ...              41          0
210          0            0  ...              52          0
352          0            0  ...             157          0
377        188            0  ...              37          0
264     180542            0  ...             187          0

[1342 rows x 7 columns]

In [ ]:
new_data_full.sample(20)

session  student_Id exercise  ... mouse_click_right mouse_movement keystroke
375        4          44   Es_4_5  ...                 0            561         0
454        3          44   Es_3_4  ...                 0            226         0
604        4          44   Es_4_2  ...                 0            218         0
339        6          30   Es_6_3  ...                 0            118         0
76         5          15   Es_5_2  ...                 0            308         0
253        2          15   Es_2_4  ...                 0            141         0
92         5          30   Es_5_1  ...                 1             39         0
176        3          15   Es_3_2  ...                 0             40         0
169        4          15   Es_4_4  ...                 0            270         0
300        3          30   Es_3_2  ...                 0             75         0
15         3          15   Es_3_1  ...                 0             95         0
107        3          30   Es_3_1  ...                 0             95         0
263        3          15   Es_3_4  ...                 0            371         0
247        1          30   Es_1_2  ...                 0            134         0
10         6          44       Es  ...                 0             46        15
292        4          30   Es_4_2  ...                 4           1909         0
19         4          30   Es_4_1  ...                 0             67         0
15         3          44   Es_3_1  ...                 0            428         0
301        1          30   Es_1_2  ...                 0             86         0
235        1          15   Es_1_2  ...                 2             45         0

[20 rows x 13 columns]

In [ ]:
#Проведем нормализацию данных, для того, чтобы каждый из признаков имел равную степень влияния
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
#Посмотрим, что получилось
X_train

array([[-0.08844306, -0.10219389,  0.        , ..., -0.23794713,
        -0.27795954, -0.1413886 ],
       [-0.06236259, -0.10219389,  0.        , ..., -0.23794713,
         0.69414887, -0.1413886 ],
       [-0.08966835,  0.23811429,  0.        , ..., -0.23794713,
         0.16172201, -0.1413886 ],
       ...,
       [-0.08968668, -0.10219389,  0.        , ...,  2.06764891,
        -0.30200462, -0.1413886 ],
       [-0.08957899, -0.10219389,  0.        , ..., -0.23794713,
        -0.4394051 , -0.1413886 ],
       [ 0.01373359, -0.10219389,  0.        , ..., -0.23794713,
        -0.2676545 , -0.1413886 ]])

## Оценка
1. Классификационная оценка точности.
2. Матрица ошибок
3. Отчет о классификации

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
from sklearn.metrics import classification_report

## 1. Логистическая регресия
Условия логистической регрессии

1. Бинарная логистическая регрессия требует, чтобы зависимая переменная также была бинарной.
2. Для бинарной регрессии фактор уровня 1 зависимой переменной должен представлять желаемый вывод.
3. Логистическая регрессия требует больших размеров выборки.
4. Независимые переменные связаны линейно с логарифмическими коэффициентами.
5. Независимые переменные должны быть независимы друг от друга. Это значит, что модель должна иметь малую мультиколлинеарность или не иметь её вовсе.
6. Использоваться должны только значимые переменные.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
# Посмотрим на значения зависимой (целевой) переменой
y[:10]

0        Other
3        Other
7        Other
18     Diagram
20     Diagram
23       Other
27     Diagram
29     Diagram
32     Diagram
33       Other
Name: activity, dtype: object

In [ ]:
#Так как целевая переменная имеет категориальное значение, то ее необходимо бинаризовать для того, чтобы применить алгоритм логистической регрессии.
new_data_full['activity'].replace({' Other': 0, ' Diagram': 1}, inplace=True)
new_data_full.head()

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4582: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


session  student_Id exercise  ...  mouse_click_right mouse_movement keystroke
0         1          15       Es  ...                  0             83         0
3         1          15   Es_1_1  ...                  2             20         0
7         1          15   Es_1_1  ...                  2             91         0
18        1          15   Es_1_1  ...                  0            375         0
20        1          15   Es_1_1  ...                  8           8485         0

[5 rows x 13 columns]

In [ ]:
# Определим бинаризованную целевую переменную
y_log = new_data_full['activity']

In [ ]:
# Запустим алгоритм на всех данных
log_model = LogisticRegression()
log_model.fit(X, y_log)

log_pred = log_model.predict(X)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_log, log_pred).round(2), '\n')
print('Confusion matrix \n', confusion_matrix(y_log, log_pred), '\n')
print('Classification report \n', classification_report(y_log, log_pred))

Accuracy Metrics for Logistic Regression: 0.36 

Confusion matrix 
 [[   9 1432]
 [   0  796]] 

Classification report 
               precision    recall  f1-score   support

           0       1.00      0.01      0.01      1441
           1       0.36      1.00      0.53       796

    accuracy                           0.36      2237
   macro avg       0.68      0.50      0.27      2237
weighted avg       0.77      0.36      0.20      2237



Видно, что наш алгоритм относит все данные к Diagram

In [ ]:
#Теперь разделим целевую переменную на тестовую и тренировочную части
y_log_train, y_log_test = train_test_split(y_log, test_size = 0.4, random_state = 1)

In [ ]:
# Запустим алгоритм на треноровочных и тестовых данных
log_model = LogisticRegression()
log_model.fit(X_train, y_train)

log_pred = log_model.predict(X_test)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_test, log_pred).round(3), '\n')
print('Confusion matrix \n', confusion_matrix(y_test, log_pred), '\n')
print('Classification report \n', classification_report(y_test, log_pred))

Accuracy Metrics for Logistic Regression: 0.712 

Confusion matrix 
 [[108 207]
 [ 51 529]] 

Classification report 
               precision    recall  f1-score   support

     Diagram       0.68      0.34      0.46       315
       Other       0.72      0.91      0.80       580

    accuracy                           0.71       895
   macro avg       0.70      0.63      0.63       895
weighted avg       0.70      0.71      0.68       895



Уже лучше, алгоритм разделяет Diagram и Other, предпочтение отдает Other

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
# Усиленные условия регулизации C (уменьшение выбросов, стандартное занчение = 1.0)
log_model = LogisticRegression(C=0.001)
log_model.fit(X_train, y_train)

log_pred = log_model.predict(X_test)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_test, log_pred).round(3), '\n')
print('Confusion matrix \n', confusion_matrix(y_test, log_pred), '\n')
print('Classification report \n', classification_report(y_test, log_pred))

Accuracy Metrics for Logistic Regression: 0.661 

Confusion matrix 
 [[ 12 303]
 [  0 580]] 

Classification report 
               precision    recall  f1-score   support

     Diagram       1.00      0.04      0.07       315
       Other       0.66      1.00      0.79       580

    accuracy                           0.66       895
   macro avg       0.83      0.52      0.43       895
weighted avg       0.78      0.66      0.54       895



Точность результата уменьшилась, скорее всего переобучения нет

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
# Повышаем терпимость алгоритма к ошибкам (стандартное значение = 0.0001)
log_model = LogisticRegression(tol=100)
log_model.fit(X_train, y_train)

log_pred = log_model.predict(X_test)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_test, log_pred).round(3), '\n')
print('Confusion matrix \n', confusion_matrix(y_test, log_pred), '\n')
print('Classification report \n', classification_report(y_test, log_pred))

Accuracy Metrics for Logistic Regression: 0.716 

Confusion matrix 
 [[ 69 246]
 [  8 572]] 

Classification report 
               precision    recall  f1-score   support

     Diagram       0.90      0.22      0.35       315
       Other       0.70      0.99      0.82       580

    accuracy                           0.72       895
   macro avg       0.80      0.60      0.59       895
weighted avg       0.77      0.72      0.65       895



При повышении толерантности, точность повысилась

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
# Изменине алгоритма вычисления с lbfgs на ‘liblinear’ и уменьшим количество итераций (стандартное значение = 100)
log_model = LogisticRegression(solver='liblinear', max_iter=1)
log_model.fit(X_train, y_train)

log_pred = log_model.predict(X_test)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_test, log_pred).round(3), '\n')
print('Confusion matrix \n', confusion_matrix(y_test, log_pred), '\n')
print('Classification report \n', classification_report(y_test, log_pred))

Accuracy Metrics for Logistic Regression: 0.708 

Confusion matrix 
 [[ 56 259]
 [  2 578]] 

Classification report 
               precision    recall  f1-score   support

     Diagram       0.97      0.18      0.30       315
       Other       0.69      1.00      0.82       580

    accuracy                           0.71       895
   macro avg       0.83      0.59      0.56       895
weighted avg       0.79      0.71      0.63       895



/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Точность понизилась, меньшее количество итераций было причиной

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
# Изменине алгоритма вычисления с lbfgs на ‘newton-cg’
log_model = LogisticRegression(solver='newton-cg')
log_model.fit(X_train, y_train)

log_pred = log_model.predict(X_test)

print('Accuracy Metrics for Logistic Regression:', accuracy_score(y_test, log_pred).round(3), '\n')
print('Confusion matrix \n', confusion_matrix(y_test, log_pred), '\n')
print('Classification report \n', classification_report(y_test, log_pred))

Accuracy Metrics for Logistic Regression: 0.712 

Confusion matrix 
 [[108 207]
 [ 51 529]] 

Classification report 
               precision    recall  f1-score   support

     Diagram       0.68      0.34      0.46       315
       Other       0.72      0.91      0.80       580

    accuracy                           0.71       895
   macro avg       0.70      0.63      0.63       895
weighted avg       0.70      0.71      0.68       895



При очередной смене алгоритма, точность не изменилась

## 2. Ближайшие соседи (K-Nearest Neigbors)

In [ ]:
# Импортируем готовый алгоритм
from sklearn.neighbors import KNeighborsClassifier

Основные параметры
*   **n_neighbors: int, default=5** - количество соседей, по которым будет производится расчет, 
*   **weights: {‘uniform’, ‘distance’} or callable, default=’uniform’**- весовая функция, используемая в прогнозировании,
*  **algorithm: {‘auto’, ‘ball_tree’, ‘kd_tree’, ‘brute’}, default=’auto’** - aлгоритм, используемый для вычисления ближайших соседей,
*   **leaf_size: int, default=30** - размер листа передается в BallTree или KDTree. Это может повлиять на скорость построения и запроса, а также на объем памяти, необходимый для хранения дерева. Оптимальное значение зависит от характера проблемы.,
*  **p:int, default=2** - степенный параметр для метрики Минковского. Когда p = 1, это эквивалентно использованию manhattan_distance (l1) и euclidean_distance (l2) для p = 2. Для произвольного p используется minkowski_distance (l_p),
*   **metric: str or callable, default=’minkowski’** - метрика расстояния, используемая для дерева. По умолчанию используется метрика Минковского, а при p = 2 она эквивалентна стандартной евклидовой метрике. См. Документацию DistanceMetric для получения списка доступных показателей. Если метрика «вычисляется заранее», предполагается, что X является матрицей расстояний и должен быть квадратным во время подбора. X может быть разреженным графом, и в этом случае только «ненулевые» элементы могут считаться соседями.
*   **metric_params: dict, default=None** - дополнительные аргументы ключевого слова для метрической функции.,
*   **n_jobs: int, default=None** - количество параллельных заданий для поиска соседей.

In [ ]:
# Запустим алгоритм на всех данных
knn_model = KNeighborsClassifier()
knn_model.fit(X, y)
knn_pred = knn_model.predict(X)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y, knn_pred), '\n')
print(classification_report(y, knn_pred))

Accuracy Metrics for KNN: 0.764 

Confusion_matrix 
 [[1264  177]
 [ 352  444]] 

              precision    recall  f1-score   support

           0       0.78      0.88      0.83      1441
           1       0.71      0.56      0.63       796

    accuracy                           0.76      2237
   macro avg       0.75      0.72      0.73      2237
weighted avg       0.76      0.76      0.76      2237



In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.722 

Confusion_matrix 
 [[163 152]
 [ 97 483]] 

              precision    recall  f1-score   support

     Diagram       0.63      0.52      0.57       315
       Other       0.76      0.83      0.80       580

    accuracy                           0.72       895
   macro avg       0.69      0.68      0.68       895
weighted avg       0.71      0.72      0.71       895



Точность при использовании тренировочных и тестовых данных ниже, чем при использовании всех данных

Попробуем это исправить

In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
#Увеличим количество соседей до 13
knn_model = KNeighborsClassifier(n_neighbors=13)
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.734 

Confusion_matrix 
 [[153 162]
 [ 76 504]] 

              precision    recall  f1-score   support

     Diagram       0.67      0.49      0.56       315
       Other       0.76      0.87      0.81       580

    accuracy                           0.73       895
   macro avg       0.71      0.68      0.69       895
weighted avg       0.73      0.73      0.72       895



При перебирании значений соседей (от 1 до 17 включительно) было выявлено, что наибольшая точность достигается при 13 соседях (0.734)

Теперь попробуем еще увеличить точность

In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
#Соседей - 16
#Изменим весовую функцию weights='uniform' на weights='distance'. Таким образом алгоритм будет больше обрашать на дистанцию соседей
knn_model = KNeighborsClassifier(n_neighbors=16, weights='distance')
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.717 

Confusion_matrix 
 [[170 145]
 [108 472]] 

              precision    recall  f1-score   support

     Diagram       0.61      0.54      0.57       315
       Other       0.76      0.81      0.79       580

    accuracy                           0.72       895
   macro avg       0.69      0.68      0.68       895
weighted avg       0.71      0.72      0.71       895



Изменение весовой функции во всех случаях дает меньшее значение точности (максимальная на 16 соседях)

Попробуем что-то другое

In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
#Соседей - 13
#Попробуем разные алгоритмы algorithm{‘auto’(0.734), ‘ball_tree’(0.734), ‘kd_tree’(0.734), ‘brute’(0.733)}, default=’auto’
knn_model = KNeighborsClassifier(n_neighbors=13, algorithm='brute')
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.733 

Confusion_matrix 
 [[153 162]
 [ 77 503]] 

              precision    recall  f1-score   support

     Diagram       0.67      0.49      0.56       315
       Other       0.76      0.87      0.81       580

    accuracy                           0.73       895
   macro avg       0.71      0.68      0.68       895
weighted avg       0.72      0.73      0.72       895



Ничего не поменялось, возьмем другое

In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
#Соседей - 13
#Попробуем изменить leaf_size : int, default=30 (При 10 повысилась точность с 0.734 до 0.735)
knn_model = KNeighborsClassifier(n_neighbors=13, leaf_size=10)
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.735 

Confusion_matrix 
 [[154 161]
 [ 76 504]] 

              precision    recall  f1-score   support

     Diagram       0.67      0.49      0.57       315
       Other       0.76      0.87      0.81       580

    accuracy                           0.74       895
   macro avg       0.71      0.68      0.69       895
weighted avg       0.73      0.74      0.72       895



Ну и последняя попытка улучшить результат

In [ ]:
#Запустим алгоритм на тренировочных и тестовых данных
#Соседей - 13
#leaf_size - 10
#Изменим количество параллельных заданий для поиска соседей. n_jobs : int, default=None (Тоже ничего не меняется)
knn_model = KNeighborsClassifier(n_neighbors=13, leaf_size=10, n_jobs=1)
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

print('Accuracy Metrics for KNN:', 
      accuracy_score(y_test, knn_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, knn_pred), '\n')
print(classification_report(y_test, knn_pred))

Accuracy Metrics for KNN: 0.735 

Confusion_matrix 
 [[154 161]
 [ 76 504]] 

              precision    recall  f1-score   support

     Diagram       0.67      0.49      0.57       315
       Other       0.76      0.87      0.81       580

    accuracy                           0.74       895
   macro avg       0.71      0.68      0.69       895
weighted avg       0.73      0.74      0.72       895



## 3. Дерево решений

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
# Запустим алгоритм на всех данных
cart_model = DecisionTreeClassifier()
cart_model.fit(X, y)
cart_pred = cart_model.predict(X)

print('Accuracy Metrics for Decision Tree Classifier:', 
      accuracy_score(y, cart_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y, cart_pred), '\n')
print(classification_report(y, cart_pred))

Accuracy Metrics for Decision Tree Classifier: 0.96 

Confusion_matrix 
 [[1434    7]
 [  83  713]] 

              precision    recall  f1-score   support

           0       0.95      1.00      0.97      1441
           1       0.99      0.90      0.94       796

    accuracy                           0.96      2237
   macro avg       0.97      0.95      0.96      2237
weighted avg       0.96      0.96      0.96      2237



In [ ]:
# Запустим алгоритм на тренировочных и тестовых данных
cart_model = DecisionTreeClassifier()
cart_model.fit(X_train, y_train)
cart_pred = cart_model.predict(X_test)

print('Accuracy Metrics for Decision Tree Classifier:', 
      accuracy_score(y_test, cart_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, cart_pred), '\n')
print(classification_report(y_test, cart_pred))

Accuracy Metrics for Decision Tree Classifier: 0.65 

Confusion_matrix 
 [[167 148]
 [165 415]] 

              precision    recall  f1-score   support

     Diagram       0.50      0.53      0.52       315
       Other       0.74      0.72      0.73       580

    accuracy                           0.65       895
   macro avg       0.62      0.62      0.62       895
weighted avg       0.65      0.65      0.65       895



При обучении на всех данных точность колласальная и равна 0.96

При обучении на тренировочных и тестовых данных точность довольно низкая (0.654), попробуем исправить

In [ ]:
# Запустим алгоритм на тренировочных и тестовых данных
#Добавим глубину max_depth=None, лучший результат при значении 3
cart_model = DecisionTreeClassifier(max_depth=3)
cart_model.fit(X_train, y_train)
cart_pred = cart_model.predict(X_test)

print('Accuracy Metrics for Decision Tree Classifier:', 
      accuracy_score(y_test, cart_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, cart_pred), '\n')
print(classification_report(y_test, cart_pred))

Accuracy Metrics for Decision Tree Classifier: 0.712 

Confusion_matrix 
 [[ 85 230]
 [ 28 552]] 

              precision    recall  f1-score   support

     Diagram       0.75      0.27      0.40       315
       Other       0.71      0.95      0.81       580

    accuracy                           0.71       895
   macro avg       0.73      0.61      0.60       895
weighted avg       0.72      0.71      0.67       895



Ограничив глубину дерева до 3, получилось улучшить алгоритм, попробуем что-то еще

In [ ]:
# Запустим алгоритм на тренировочных и тестовых данных
#max_depth=None
#Изменим min_samples_split=2 на что-нибудь другое
cart_model = DecisionTreeClassifier(max_depth=None, min_samples_split=50)
cart_model.fit(X_train, y_train)
cart_pred = cart_model.predict(X_test)

print('Accuracy Metrics for Decision Tree Classifier:', 
      accuracy_score(y_test, cart_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, cart_pred), '\n')
print(classification_report(y_test, cart_pred))

Accuracy Metrics for Decision Tree Classifier: 0.717 

Confusion_matrix 
 [[161 154]
 [ 99 481]] 

              precision    recall  f1-score   support

     Diagram       0.62      0.51      0.56       315
       Other       0.76      0.83      0.79       580

    accuracy                           0.72       895
   macro avg       0.69      0.67      0.68       895
weighted avg       0.71      0.72      0.71       895



При деление на 50 минимальных выборок, основной узел делится таким образом, что условная точность возрастает до 0.717

In [ ]:
# Запустим алгоритм на тренировочных и тестовых данных
#max_leaf_nodes=None, изменим данный параметр
cart_model = DecisionTreeClassifier(max_leaf_nodes=47)
cart_model.fit(X_train, y_train)
cart_pred = cart_model.predict(X_test)

print('Accuracy Metrics for Decision Tree Classifier:', 
      accuracy_score(y_test, cart_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, cart_pred), '\n')
print(classification_report(y_test, cart_pred))

Accuracy Metrics for Decision Tree Classifier: 0.741 

Confusion_matrix 
 [[172 143]
 [ 89 491]] 

              precision    recall  f1-score   support

     Diagram       0.66      0.55      0.60       315
       Other       0.77      0.85      0.81       580

    accuracy                           0.74       895
   macro avg       0.72      0.70      0.70       895
weighted avg       0.73      0.74      0.73       895



Получилось довести точность до 0.741 ограничив максимальное количество узлов, которые выбирались по принципу лучший первый, с наименьшим количеством примесей 

## Случайный лес

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
# Запустим алгоритм на всех данных
rfc_model = RandomForestClassifier()
rfc_model.fit(X, y)
rfc_pred = rfc_model.predict(X)

print('Accuracy Metrics for Random Forest Classifier:', 
      accuracy_score(y, rfc_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y, rfc_pred), '\n')
print(classification_report(y, rfc_pred))

Accuracy Metrics for Random Forest Classifier: 0.96 

Confusion_matrix 
 [[1418   23]
 [  67  729]] 

              precision    recall  f1-score   support

           0       0.95      0.98      0.97      1441
           1       0.97      0.92      0.94       796

    accuracy                           0.96      2237
   macro avg       0.96      0.95      0.96      2237
weighted avg       0.96      0.96      0.96      2237



In [ ]:
# Запустим алгоритм на обучающих и тестовых данных

rfc_model = RandomForestClassifier()
rfc_model.fit(X_train, y_train)
rfc_pred = rfc_model.predict(X_test)

print('Accuracy Metrics for Random Forest Classifier:', 
      accuracy_score(y_test, rfc_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, rfc_pred), '\n')
print(classification_report(y_test, rfc_pred))

Accuracy Metrics for Random Forest Classifier: 0.713 

Confusion_matrix 
 [[174 141]
 [116 464]] 

              precision    recall  f1-score   support

     Diagram       0.60      0.55      0.58       315
       Other       0.77      0.80      0.78       580

    accuracy                           0.71       895
   macro avg       0.68      0.68      0.68       895
weighted avg       0.71      0.71      0.71       895



Результат удовлетворительный, попробуем улучшить его

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
#n_estimators=100, изменим количество деревьев в лесу
rfc_model = RandomForestClassifier(n_estimators=114)
rfc_model.fit(X_train, y_train)
rfc_pred = rfc_model.predict(X_test)

print('Accuracy Metrics for Random Forest Classifier:', 
      accuracy_score(y_test, rfc_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, rfc_pred), '\n')
print(classification_report(y_test, rfc_pred))

Accuracy Metrics for Random Forest Classifier: 0.717 

Confusion_matrix 
 [[176 139]
 [114 466]] 

              precision    recall  f1-score   support

     Diagram       0.61      0.56      0.58       315
       Other       0.77      0.80      0.79       580

    accuracy                           0.72       895
   macro avg       0.69      0.68      0.68       895
weighted avg       0.71      0.72      0.71       895



Увеличив количество деревьев до 114, получилось добиться дополнительных 0.004 точности

Попробуем что-то другое

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
#Изменим criterion=“gini”, на “entropy”, функцию измерения качества деления деревьев
rfc_model = RandomForestClassifier(criterion='entropy')
rfc_model.fit(X_train, y_train)
rfc_pred = rfc_model.predict(X_test)

print('Accuracy Metrics for Random Forest Classifier:', 
      accuracy_score(y_test, rfc_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, rfc_pred), '\n')
print(classification_report(y_test, rfc_pred))

Accuracy Metrics for Random Forest Classifier: 0.721 

Confusion_matrix 
 [[178 137]
 [113 467]] 

              precision    recall  f1-score   support

     Diagram       0.61      0.57      0.59       315
       Other       0.77      0.81      0.79       580

    accuracy                           0.72       895
   macro avg       0.69      0.69      0.69       895
weighted avg       0.72      0.72      0.72       895



Аккуратность модели увеличилось

Попробуем что-то еще

In [ ]:
# Запустим алгоритм на обучающих и тестовых данных
#Изменим ccp_alpha=0.0, Параметр сложности, используемый для обрезки с минимальными затратами. 
#Будет выбрано поддерево с наибольшей сложностью затрат, меньшее, чем ccp_alpha. По умолчанию обрезка не выполняется.
rfc_model = RandomForestClassifier(ccp_alpha=0.001)
rfc_model.fit(X_train, y_train)
rfc_pred = rfc_model.predict(X_test)

print('Accuracy Metrics for Random Forest Classifier:', 
      accuracy_score(y_test, rfc_pred).round(3), '\n')
print('Confusion_matrix \n', 
      confusion_matrix(y_test, rfc_pred), '\n')
print(classification_report(y_test, rfc_pred))

Accuracy Metrics for Random Forest Classifier: 0.731 

Confusion_matrix 
 [[172 143]
 [ 98 482]] 

              precision    recall  f1-score   support

     Diagram       0.64      0.55      0.59       315
       Other       0.77      0.83      0.80       580

    accuracy                           0.73       895
   macro avg       0.70      0.69      0.69       895
weighted avg       0.72      0.73      0.73       895



Получилось добиться результата точности равный 0.731